### PR TITLE
Move away from global state

### PR DESCRIPTION
**Pull request by fichilleen**
Removed global state from most places to allow nicer unit testing and generally lower dependency on the rest of the game being up

Stubbed out the basics of some tests here too. Removed some nice code, and added some ugly code, so it's not perfect, but a step in the right direction

### CODE REVIEW

File: lib/command.py [hunk 1]
```diff
@@ -1,108 +1,113 @@
...
+            "look": self.look,
+            "go": self.go,
+        }
+
+    def execute_command(self, player, command, param):
+        print(self.commands)
+        if self.commands.get(command):
+            self.commands[command](player, param)
+        else:
+            self.game_state.tell_player(player, "Unknown command '{}'".format(command))
```
Comment (fado) [inline lib/command.py:28]: Would much prefer this to be `tell_player()` without the need to go down through all these layers of objects. I don't know if it makes sense having the `tell_player()` method on a `GameState` object.

File: lib/models/player.py [hunk 2]
```diff
@@ -1,10 +1,10 @@
-from dataclasses import dataclass
+from abc import ABC
 
 from lib.models.client import Client
 from lib.models.creature import Creature
 
 
-class Player(Creature):
+class Player(Creature, ABC):
```
Comment (fado) [inline lib/models/player.py:7]: What's the purpose of this?

File: mudserver.py [hunk 3]
```diff
@@ -149,6 +149,7 @@ def _check_for_new_connections(self):
             self._new_events.append(
                 Event(ServerEvents.NEW_PLAYER, client)
             )
+            print(self._new_events)
```
Comment (fado) [inline mudserver.py:152]: We should probably get logging going soon.

File: lib/models/game_state.py [hunk 4]
```diff
@@ -3,8 +3,11 @@
 from mudserver import MudServer
 from lib.models.player import Player
 
+# TODO: Is there too much in this class now?
```
Comment (fado) [inline lib/models/game_state.py:6]: I'm just not keen on how the methods here are reached now.

File: lib/command.py [hunk 5]
```diff
@@ -1,108 +1,113 @@
...
-
-    # the specified exit wasn't found in the current room
-    else:
-        # send back an 'unknown exit' message
-        tell_player(player, f"Unknown exit '{ex}'")
+        # if the specified exit is found in the room's exits list
+        if ex in current_location["exits"]:
+
+            # go through all the players in the game
+            for other_player in self.game_state.list_players():
```
Comment (fado) [inline lib/command.py:85]: Again, I would love if there was some way we could more cleanly expose these methods.

**Review by fado — APPROVED** [review]
Tentative approval, but not thrilled about the interface changes.

File: lib/models/player.py (hunk 2 above)
Comment (fichilleen) [reply lib/models/player.py:7]: Mocking multiple inheritance

File: mudserver.py (hunk 3 above)
Comment (fichilleen) [reply mudserver.py:152]: Aye didn't mean to leave that in really

**Comment by fichilleen** [comment]
The big thing is, we either need to decide to put these as classless methods that take a pile of arguments, and those arguments are passed around everywhere, or embed them in classes some way or another. I don't have a strong opinion on which way to do it, I just did it this way to get the state out of the main namespace

File: lib/command.py (hunk 1 above)
Comment (fichilleen) [reply lib/command.py:28]: The problem is, tell player needs to know about the same state somehow, so we can either have:

`GameState.tell_player(player, message)`

`tell_player(gamestate, player, message)`

`Player.tell(message)` <- with this one, we still needed to find the Player, so there's an implicit call to gamestate somewhere

**Pull request merged**